scripts/publish.js

```diff
@@ -0,0 +1,442 @@
+ * Default path to workspace root
+ */
+const DEFAULT_ROOT = path.join(__dirname, '..')
+
+/**
+ * Default options provided to a {@link GlobFn}
+ */
+const DEFAULT_GLOB_OPTS = /** @type {const} */ ({
+  cwd: DEFAULT_ROOT,
+  withFileTypes: true,
+  ignore: {
+    ignored: /** @param {GlobDirent} p */ (p) => !p.parent || !p.isDirectory(),
+  },
+})
+
+/**
+ * Default options provided to a {@link SpawnFn}
+ */
+const DEFAULT_SPAWN_OPTS = /** @type {const} */ ({
+  cwd: DEFAULT_ROOT,
+  stdio: 'inherit',
+  shell: true,
+})
+
+/**
+ * `Dirent`-like object returned by `glob`; adds a `fullpath()` method and `parent` prop
+ *
+ * @defaultValue `Path` from `path-scurry`, which is resolved by `glob()` if the
+ * `withFileTypes` option is `true`
+ * @typedef {import('fs').Dirent & {fullpath: () => string, parent?: GlobDirent}} GlobDirent
+ */
+
+/**
+ * Function used to spawn `npm publish`.
+ *
+ * Returned `EventEmitter` _must_ emit `exit` and _may_ emit `error`.
+ *
+ * @defaultValue `childProcess.spawn`
+ * @typedef {(cmd: string, args: string[], opts: typeof DEFAULT_SPAWN_OPTS) => import('events').EventEmitter} SpawnFn
+ */
+
+/**
+ * Function used to execute commands and retrieve the `stdout` and `stderr` from
+ * execution.
+ *
+ * @typedef {(cmd: string, opts: {cwd: string}) => Promise<{stdout: string, stderr: string}>} ExecFn
+ */
+
+/**
+ * Globbing function.
+ *
+ * Pretty tightly-bound to `glob`, unfortunately.
+ *
+ * @defaultValue `Glob.glob`
+ * @typedef {(pattern: string | string[], opts: typeof DEFAULT_GLOB_OPTS) =>
+ * Promise<GlobDirent[]>} GlobFn
+ */
+
+/**
+ * Function used to parse a JSON string
+ *
+ * @defaultValue `JSON.parse`
+ * @typedef {(json: string) => any} ParseJsonFn
+ */
+
+/**
+ * Loosey-gooesy `fs.promises` implementation
+ * @typedef {{[K in keyof Pick<typeof Fs.promises, 'lstat' | 'readdir' | 'readlink' | 'realpath' | 'readFile'>]: (...args: any[]) => Promise<any>}} MinimalFsPromises
+ */
+
+/**
+ * Bare minimum `fs` implementation for our purposes
+ * @typedef MinimalFs
+ * @property {MinimalFsPromises} promises
+ */
+
+/**
+ * Options for {@link publishWorkspaces}
+ *
+ * @typedef PublishWorkspacesOptions
+ * @property {boolean} [dryRun] - Whether to publish in dry-run mode
+ * @property {MinimalFs} [fs] - Bare minimum `fs` implementation
+ * @property {GlobFn} [glob] - Function to glob for files
+ * @property {ExecFn} [exec] - Function to execute a command
+ * @property {SpawnFn} [spawn] - Function to spawn a process
+ * @property {ParseJsonFn} [parseJson] - Function to parse JSON
+ * @property {Console} [console] - Console to use for logging
+ * @property {string} [root] - Workspace root
+ * @property {string[]} [newPkg] - New packages to publish
+ */
+
+/**
+ * Minimal `package.json` definition.
+ *
+ * @todo If we ever pull in `type-fest` for whatever reason, use its `PackageJson` instead.
+ * @typedef PackageJson
+ * @property {boolean} [private]
+ * @property {string} name
+ * @property {string} version
+ */
```

Comment:
(could) This might work better if put in a d.ts file

Reply:
will look at that if/when we end up transitioning this to a package.